### PR TITLE
[UIE-16] Move DelayedRender to common components

### PR DESCRIPTION
**Pull request by nawatts**
Move DelayedRender to common components, convert to TypeScript and add tests.

### CODE REVIEW

File: src/components/icons.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import _ from 'lodash/fp'
 import { Children, Fragment } from 'react'
 import { div, h, span } from 'react-hyperscript-helpers'
-import DelayedRender from 'src/components/DelayedRender'
+// Temporary workaround to avoid a circular import.
```
Comment (nawatts) [inline src/components/icons.js:4]: Didn't want to deal with this here.

File: src/components/common/DelayedRender.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+}>
+
+export const DelayedRender = ({ children = null, delay = 1000 }: DelayedRenderProps) => {
+  const [shouldRender, setShouldRender] = useState(false)
+  useEffect(() => {
+    const timeout = setTimeout(() => setShouldRender(true), delay)
+    return () => clearTimeout(timeout)
+  }, [delay])
+
+  return shouldRender ? h(Fragment, [children]) : null
```
Comment (nawatts) [inline src/components/common/DelayedRender.ts:16]: Had to change this line from `shouldRender && children` to `shouldRender ? h(Fragment, [children]) : null` to make the types for react-hyperscript-helpers happy.

**Review by marctalbott — APPROVED** [review]

File: src/components/common/DelayedRender.test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    // Arrange
+    jest.useFakeTimers()
+
+    // Act
+    render(h(DelayedRender, { delay: 3000 }, [
+      div(['Hello world'])
+    ]))
+
+    // Assert
+    expect(screen.queryByText('Hello world')).toBeNull()
```
Comment (msilva-broad) [inline src/components/common/DelayedRender.test.ts]: this and other intermediate assertions can be switched to saved values to be asserted later in a unified // Assert section.

```
const initialIsContentRendered = screen.queryByText('Hello world') === null
// ...
// Assert
expect(initialIsContentRendered).toBe(false)
```

Then the // Act section can be larger to encompass the multi-step flow being tested.

File: src/components/common/DelayedRender.ts (hunk 2 above)
Comment (msilva-broad) [reply src/components/common/DelayedRender.ts:16]: I wonder if a maybeRender(condition: boolean, ...) helper would be good to automatically return the null on false?

File: src/components/common/DelayedRender.test.ts (hunk 3 above)
Comment (nawatts) [reply src/components/common/DelayedRender.test.ts]: Done in https://github.com/DataBiosphere/terra-ui/pull/3795/commits/39369d085bfd9bca37e0bfb661f8757a85cb00f9. I still find switching between act and assert more convenient and sometimes easier to read than storing intermediate values to keep all assertions at the end.

**Review by msilva-broad — APPROVED** [review]

File: src/components/common/DelayedRender.ts (hunk 2 above)
Comment (nawatts) [reply src/components/common/DelayedRender.ts:16]: I'm not sure that's necessary. React is fine with returning false here, it's just our react-hyperscript-helpers types that are unhappy with it.

**Pull request merged**